File: examples/trello_github/orchestrator.py
"""Executable Trello + GitHub + Claude Code composition."""

from __future__ import annotations

import argparse
import os
from pathlib import Path

from orchestra.channels import ConsoleChannel, HealthChangeChannel, JsonlAuditChannel
from orchestra.engine import Engine, EngineConfig
from orchestra.runtime import (
    ClaudeCodeRuntime,
    ClaudeRuntimeConfig,
    CodexCliRuntime,
    CodexRuntimeConfig,
)
from orchestra.state import FileStateStore, TickAlreadyRunning

from .adapters import GitHubActivitySource, TrelloBoard

ROOT = Path(__file__).resolve().parents[2]


def _integer(name: str, default: int) -> int:
    try:
        return max(0, int(os.environ.get(name, str(default))))
    except ValueError:
        return default


def build_engine(*, dry_run: bool = False) -> Engine:
    ready_lists = tuple(
        value.strip()
        for value in os.environ.get("TRELLO_READY_LIST_IDS", "").split(",")
        if value.strip()
    )
    passthrough = tuple(
        value.strip()
        for value in os.environ.get("ORCHESTRA_AGENT_ENV", "").split(",")
        if value.strip()
    )
    store = FileStateStore(ROOT / "var", ROOT / "templates" / "task")
    board = TrelloBoard(
        key=os.environ.get("TRELLO_KEY", ""),
        token=os.environ.get("TRELLO_TOKEN", ""),
        board_id=os.environ.get("TRELLO_BOARD_ID", ""),
        ready_list_ids=ready_lists,
    )
    github = GitHubActivitySource(
        token=os.environ.get("GITHUB_TOKEN", ""),
        bot_login=os.environ.get("GITHUB_BOT_LOGIN", ""),
        api_url=os.environ.get("GITHUB_API_URL", "https://api.github.com"),
    )
    engine_name = os.environ.get("ORCHESTRA_AGENT_ENGINE", "claude").lower()
    if engine_name == "claude":
        runtime = ClaudeCodeRuntime(
            store,
            ClaudeRuntimeConfig(
                executable=os.environ.get("CLAUDE_BIN", "claude"),
                model=os.environ.get("ORCHESTRA_MODEL", ""),
                permission_mode=os.environ.get(
                    "ORCHESTRA_PERMISSION_MODE", "default"
                ),
                passthrough_env=passthrough,
            ),
        )
    elif engine_name == "codex":
        runtime = CodexCliRuntime(
            store,
            CodexRuntimeConfig(
                executable=os.environ.get("CODEX_BIN", "codex"),
                model=os.environ.get("ORCHESTRA_MODEL", ""),
                sandbox=os.environ.get(
                    "ORCHESTRA_CODEX_SANDBOX", "workspace-write"
                ),
                network_access=(
                    os.environ.get("ORCHESTRA_CODEX_NETWORK_ACCESS", "1") == "1"
                ),
                dangerously_bypass_approvals_and_sandbox=(
                    os.environ.get("ORCHESTRA_CODEX_DANGEROUS_BYPASS") == "1"
                ),
                passthrough_env=passthrough,
            ),
        )
    else:
        raise ValueError(
            "ORCHESTRA_AGENT_ENGINE must be 'claude' or 'codex', "
            f"not {engine_name!r}"
        )
    channels = [
        ConsoleChannel(),
        JsonlAuditChannel(ROOT / "var" / "logs" / "ticks.jsonl"),
        HealthChangeChannel(store.cache_dir / "health.json"),
    ]
    return Engine(
        board=board,
        activity_sources=[github],
        runtime=runtime,
        channels=channels,
        store=store,
        config=EngineConfig(
            max_concurrent_turns=_integer("ORCHESTRA_MAX_TURNS", 3),
            stall_after_seconds=_integer("ORCHESTRA_STALL_AFTER", 900),
            max_restarts=_integer("ORCHESTRA_MAX_RESTARTS", 5),
            require_directed_activity=(
                os.environ.get("ORCHESTRA_REQUIRE_DIRECTED", "1") == "1"
            ),
            bot_identities=frozenset(
                value
                for value in [os.environ.get("GITHUB_BOT_LOGIN", "")]
                if value
            ),
            dry_run=dry_run,
        ),
    )


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--status", action="store_true")
    args = parser.parse_args()
    engine = build_engine(dry_run=args.dry_run)
    if args.status:
        for key, snapshot in sorted(engine.status().items()):
            print(f"{key}\t{snapshot.state}")
        return 0
    try:
        engine.tick()
    except TickAlreadyRunning as error:
        print(error)
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/trello_github/orchestrator.ts
#!/usr/bin/env node
/** Executable Trello + GitHub composition. One invocation performs one tick. */
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";
import { realpathSync } from "node:fs";
import {
  ConsoleChannel,
  HealthChangeChannel,
  JsonlAuditChannel,
} from "../../patchwake/channels.js";
import { Engine } from "../../patchwake/engine.js";
import { ClaudeCodeRuntime, CodexCliRuntime } from "../../patchwake/runtime.js";
import type { CodexRuntimeConfig } from "../../patchwake/runtime.js";
import type { AgentRuntime } from "../../patchwake/ports.js";
import { FileStateStore, TickAlreadyRunning } from "../../patchwake/state.js";
import { errorMessage } from "../../patchwake/json.js";
import { GitHubActivitySource, TrelloBoard } from "./adapters.js";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "../../..");
function integer(
  env: NodeJS.ProcessEnv,
  name: string,
  fallback: number,
): number {
  const text = env[name];
  return text !== undefined &&
    /^[+-]?\d+$/.test(text.trim()) &&
    Number.isSafeInteger(Number(text))
    ? Math.max(0, Number(text))
    : fallback;
}
const list = (value: string | undefined): string[] =>
  (value ?? "")
    .split(",")
    .map((part) => part.trim())
    .filter(Boolean);
export function buildEngine({
  dryRun = false,
  root = ROOT,
  env = process.env,
}: {
  dryRun?: boolean;
  root?: string;
  env?: NodeJS.ProcessEnv;
} = {}): Engine {
  const store = new FileStateStore(
    join(root, "var"),
    join(root, "templates", "task"),
  );
  const board = new TrelloBoard({
    key: env.TRELLO_KEY ?? "",
    token: env.TRELLO_TOKEN ?? "",
    boardId: env.TRELLO_BOARD_ID ?? "",
    readyListIds: list(env.TRELLO_READY_LIST_IDS),
  });
  const github = new GitHubActivitySource({
    token: env.GITHUB_TOKEN ?? "",
    botLogin: env.GITHUB_BOT_LOGIN ?? "",
    apiUrl: env.GITHUB_API_URL ?? "https://api.github.com",
  });
  const engineName = (env.PATCHWAKE_AGENT_ENGINE ?? "claude").toLowerCase();
  const common = {
    model: env.PATCHWAKE_MODEL ?? "",
    passthroughEnv: list(env.PATCHWAKE_AGENT_ENV),
  };
  let runtime: AgentRuntime;
  if (engineName === "claude")
    runtime = new ClaudeCodeRuntime(store, {
      ...common,
      executable: env.CLAUDE_BIN ?? "claude",
      permissionMode: env.PATCHWAKE_PERMISSION_MODE ?? "default",
    });
  else if (engineName === "codex") {
    const sandbox = env.PATCHWAKE_CODEX_SANDBOX ?? "workspace-write";
    if (
      sandbox !== "read-only" &&
      sandbox !== "workspace-write" &&
      sandbox !== "danger-full-access"
    )
      throw new Error(`invalid Codex sandbox: ${sandbox}`);
    const config: CodexRuntimeConfig = {
      ...common,
      executable: env.CODEX_BIN ?? "codex",
      sandbox,
      networkAccess: (env.PATCHWAKE_CODEX_NETWORK_ACCESS ?? "1") === "1",
      dangerouslyBypassApprovalsAndSandbox:
        env.PATCHWAKE_CODEX_DANGEROUS_BYPASS === "1",
    };
    runtime = new CodexCliRuntime(store, config);
  } else
    throw new Error(
      `PATCHWAKE_AGENT_ENGINE must be 'claude' or 'codex', not '${engineName}'`,
    );
  return new Engine({
    board,
    activitySources: [github],
    runtime,
    store,
    channels: [
      new ConsoleChannel(),
      new JsonlAuditChannel(join(root, "var", "logs", "ticks.jsonl")),
      new HealthChangeChannel(join(store.cacheDir, "health.json")),
    ],
    config: {
      maxConcurrentTurns: integer(env, "PATCHWAKE_MAX_TURNS", 3),
      stallAfterSeconds: integer(env, "PATCHWAKE_STALL_AFTER", 900),
      maxRestarts: integer(env, "PATCHWAKE_MAX_RESTARTS", 5),
      requireDirectedActivity: (env.PATCHWAKE_REQUIRE_DIRECTED ?? "1") === "1",
      botIdentities: new Set([env.GITHUB_BOT_LOGIN ?? ""].filter(Boolean)),
      dryRun,
    },
  });
}
export async function main(args = process.argv.slice(2)): Promise<number> {
  const { values } = parseArgs({
    args,
    options: {
      "dry-run": { type: "boolean" },
      status: { type: "boolean" },
      help: { type: "boolean", short: "h" },
    },
  });
  if (values.help) {
    console.log(
      "Usage: patchwake-trello-github [--dry-run] [--status]\nRun one Trello + GitHub orchestration tick.",
    );
    return 0;
  }
  const engine = buildEngine({ dryRun: values["dry-run"] ?? false });
  if (values.status) {
    for (const [key, snapshot] of Object.entries(await engine.status()).sort(
      ([a], [b]) => (a < b ? -1 : 1),
    ))
      console.log(`${key}\t${snapshot.state}`);
    return 0;
  }
  try {
    await engine.tick();
  } catch (error) {
    if (error instanceof TickAlreadyRunning) {
      console.log(error.message);
      return 2;
    }
    throw error;
  }
  return 0;
}
if (
  process.argv[1] &&
  realpathSync(process.argv[1]) === fileURLToPath(import.meta.url)
) {
  main()
    .then((code) => {
      process.exitCode = code;
    })
    .catch((error) => {
      console.error(errorMessage(error));
      process.exitCode = 1;
    });
}
